solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/FieldValueFeature.java

```diff
@@ -106,18 +106,29 @@ public FieldValueFeatureWeight(IndexSearcher searcher,
 
     /**
      * Return a FeatureScorer that uses docValues or storedFields if no docValues are present
+     *
      * @param context the segment this FeatureScorer is working with
      * @return FeatureScorer for the current segment and field
      * @throws IOException as defined by abstract class Feature
      */
     @Override
     public FeatureScorer scorer(LeafReaderContext context) throws IOException {
       if (schemaField != null && !schemaField.stored() && schemaField.hasDocValues()) {
-        return new DocValuesFieldValueFeatureScorer(this, context,
-            DocIdSetIterator.all(DocIdSetIterator.NO_MORE_DOCS), schemaField.getType());
+
+        FieldInfo fieldInfo = context.reader().getFieldInfos().fieldInfo(field);
+        DocValuesType docValuesType = fieldInfo != null ? fieldInfo.getDocValuesType() : DocValuesType.NONE;
+
+        if (DocValuesType.NUMERIC.equals(docValuesType) || DocValuesType.SORTED.equals(docValuesType)) {
+          return new DocValuesFieldValueFeatureScorer(this, context,
+                  DocIdSetIterator.all(DocIdSetIterator.NO_MORE_DOCS), schemaField.getType(), docValuesType);
+          // If type is NONE, this segment has no docs with this field. That's not a problem, because we won't call score() anyway
```

Comment:
Earlier today I had a thought on this code area here and wanted to note it down (for further pondering next week) but had to repair my bicycle's puncture first. Anyway ...

The _"If type is NONE, this segment has no docs with this field. That's not a problem ..."_ comment here is very helpful. I wonder though if the _"because we won't call score() anyway"_ part remains accurate after the refactorings. In the branch's initial commit I think it was accurate i.e. via the `if (docValues != null` at the start of the `score()` implementation the inside of the `if` block wasn't called. But now there's something more interesting going on here:

* If there's a segment whose documents all don't have the field then instead of a `DocValuesFieldValueFeatureScorer` the `FieldValueFeatureScorer` is used via the fall-through.
* Conceptually that jumps out as untidy or unintuitive i.e. if there is no field for all the documents in the segment then (say) a `NoFieldValueFeatureScorer` is more obvious.
* In terms of performance use of the `FieldValueFeatureScorer` should be more costly i.e. reading of the field would be attempted and nothing would be found.
* In terms of test coverage this would not be caught since the default value is returned, albeit inefficiently, by the `FieldValueFeatureScorer` implementation.
* In terms of use cases this code path may or may not be an edge case e.g. if a `FieldValueFeature` is for a field that is sparsely populated then a good proportion of segments might encounter the code path and reasons for sparseness may include (say) historical documents not having the field.

_If_ that reasoning makes sense then a next step here could be to create and use a `NoFieldValueFeatureScorer` class. What do you think?

Reply:
Great observation! I completely overlooked this.
Your reasoning makes complete sense to me and some debugging verified the behavior you described.

Even if it may be an edge case, I think it makes sense to make it explicit.
In terms of performance you are also right. Trying to read the stored fields before returning the default value surely is more costly than just returning the value.

I started to create a `DefaultValueFieldValueFeatureScorer` that always returns the default value, but until now I am struggling to find a way to test this.

I hope your bike is well now! :)

Reply:
> ... to find a way to test this. ...

Thinking conceptually (without actually trying it) one way could be like this ...
* imagine `TestFieldValueFeature` had a static inner class `FooBarFieldValueFeature extends FieldValueFeature` and for some of its tests it uses that instead of the 'real' `FieldValueFeature` class
* imagine `FooBarFieldValueFeature` can observe whatever needs to be observed and capture the observations in class level variables
* if the class level variables are public or package scope the test has access to them

Well, that's the theory but in practice what we wish to observe and test is not directly in `FieldValueFeature` but inside the `FieldValueFeatureWeight` created, but still it could be possible e.g. something like

```
class FooBarFieldValueFeature extends FieldValueFeature {

  static AtomicBoolean seenFieldValueFeatureScorer = false; // could also be counter instead of boolean

  @Override
  public FeatureWeight createWeight(...) throws IOException {
    return new FooBarFieldValueFeatureWeight(...);
  }

  public class FooBarFieldValueFeatureWeight extends FieldValueFeatureWeight {
    @Override
    public FeatureScorer scorer(LeafReaderContext context) throws IOException {
      FeatureScorer featureScorer = super.scorer();
      if (featureScorer instanceof FieldValueFeatureScorer) {
        seenFieldValueFeatureScorer = true;
      }
      return featureScorer;
    }
  }
}
```

structurally and then in the test `assertFalse(FooBarFieldValueFeature.seenFieldValueFeatureScorer);`

Does that kind of convey the idea? Essentially the idea is to extend the "real" class with a "test" class, for the test class to delegate to the real class and for the test class to be "shaped" to suit the test's purposes.

Reply:
That's a great idea!
I wanted to include the test in the `TestLTROnSolrCloud`, but then I needed to control which documents are in which segment to be able to (not) trigger the fallback. That didn't work.

Reply:
This was implemented in da6a635 with the `ObservingFieldValueFeature`.